docs/low-code-spark/extensibility/gem-builder.md

```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
+
+:::caution
+
+The content of these `Properties` classes is persisted in JSON and stored in Git, so be careful not to keep any information in them that could be a potential security risk.
+
+:::
+
+These properties are available in `validate`, `onChange` and `apply` and can be set from `dialog`, functions.
+
+````mdx-code-block
+<Tabs>
+
+<TabItem value="py" label="Python">
+
+```py
+@dataclass(frozen=True)
+    class FilterProperties(ComponentProperties):
+        columnsSelector: List[str] = field(default_factory=list)
+        condition: SColumn = SColumn("lit(True)")
+```
+</TabItem>
+
+<TabItem value="scala" label="Scala">
+
+```scala
+object Filter extends ComponentSpec {
+  val name: String = "Filter"
+  val category: String = "Transform"
+  override def optimizeCode: Boolean = true
+```
+</TabItem>
+</Tabs>
+
+````
+
+Let's look at each of these functions.
+
+### Dialog
+
+The `dialog` function contains code specific to how the Gem UI should look to the user.
+
+````mdx-code-block
+<Tabs>
+
+<TabItem value="py" label="Python">
+
+```py
+def dialog(self) -> Dialog:
+        return Dialog("Filter").addElement(
+            ColumnsLayout(height="100%")
+                .addColumn(PortSchemaTabs(selectedFieldsProperty=("columnsSelector")).importSchema(), "2fr")
+                .addColumn(StackLayout(height=("100%"))
+                .addElement(TitleElement("Filter Condition"))
+                .addElement(
+                Editor(height=("100%")).withSchemaSuggestions().bindProperty("condition.expression")
+            ), "5fr"))
+```
+</TabItem>
+<TabItem value="scala" label="Scala">
+
+```scala
+def dialog: Dialog = Dialog("Filter")
+    .addElement(
+      ColumnsLayout(height = Some("100%"))
+        .addColumn(
+          PortSchemaTabs(selectedFieldsProperty = Some("columnsSelector")).importSchema(),
+          "2fr"
+        )
+        .addColumn(
+          StackLayout(height = Some("100%"))
+            .addElement(TitleElement("Filter Condition"))
+            .addElement(
+              Editor(height = Some("100%"))
+                .withSchemaSuggestions()
+                .bindProperty("condition.expression")
+            ),
+          "5fr"
+        )
+    )
+```
+</TabItem>
+</Tabs>
+
+````
+
+The above Dialog code in the filter is rendered on UI like this
+![Dialog](img/gem-builder-ui.png)
+
+There are various UI components that can be defined for your custom Gems such as scroll boxes, tabs, buttons, and more! These UI components can be grouped together in various types of panels to create a custom user experience when using the Gem.
+
+After the Dialog object is defined, it's serialized as JSON, sent to the UI, and rendered there.
 
-Please [contact us](https://www.prophecy.io/request-a-demo) to learn more about the Gem builder.
+Depending on what kind of Gem you're creating, you'll have to define either a `Dialog` or a `DatasetDialog`.
```

Comment:
either a `Transformation Dialog` or a `Dataset Dialog`

Reply:
This is the name of the actual class we use in code. May be later we can update this if we change class names.